api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/BaseCohortDefinitionLibrary.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * The contents of this file are subject to the OpenMRS Public License
+ * Version 1.0 (the "License"); you may not use this file except in
+ * compliance with the License. You may obtain a copy of the License at
+ * http://license.openmrs.org
+ *
+ * Software distributed under the License is distributed on an "AS IS"
+ * basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See the
+ * License for the specific language governing rights and limitations
+ * under the License.
+ *
+ * Copyright (C) OpenMRS, LLC.  All Rights Reserved.
+ */
+package org.openmrs.module.eptsreports.reporting.library.cohorts;
+
+import org.openmrs.module.reporting.cohort.definition.CohortDefinition;
+import org.openmrs.module.reporting.definition.library.BaseDefinitionLibrary;
+import org.openmrs.module.reporting.definition.library.DocumentedDefinition;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+/**
+ * Defines all of the General Cohort Definition instances we want to expose for EPTS
+ */
+@Component("eptsBaseCohortDefinitionLibrary")
+public class BaseCohortDefinitionLibrary extends BaseDefinitionLibrary<CohortDefinition> {
```

Comment:
No, it's not for creating custom cohort definitions. Notice it extends [org.openmrs.module.reporting.definition.library.BaseDefinitionLibrary](https://github.com/openmrs/openmrs-module-reporting/blob/master/api/src/main/java/org/openmrs/module/reporting/definition/library/BaseDefinitionLibrary.java) and not [org.openmrs.module.reporting.cohort.definition.BaseCohortDefinition](https://github.com/openmrs/openmrs-module-reporting/blob/master/api/src/main/java/org/openmrs/module/reporting/cohort/definition/BaseCohortDefinition.java)

We are going to have a few cohort definition libraries (HivCohortDefinitionLibrary, TbCohortDefinitionLibrary, ...) and having this BaseCohortDefinitionLibrary is going to help in extracting common cohort definitions in one file/class. Extending **org.openmrs.module.reporting.definition.library.BaseDefinitionLibrary** seems to be the right thing to do and it will also allow us to document our definitions as it provides the @DocumentedDefinition annotation and #getKeyPrefix() method.

Reply:
Ok so as discussed there is a good reason for inheriting from the BaseDefinitionLibrary class. We agreed to move the age and gender queries from DataFactory and the base class into their own classes which will be referenced by the HivCohortDefinitionLibrary definition as needed. This will allow them to be reused by other classes such as dimensions if needed. Also conforms to the Single Responsiblity Principle.

Reply:
@rubailly can you please schedule some time next week to present some use cases of using the BaseDefinitionLibrary to help me better understand how it works?